internal/service/alarms/api/server.go

```diff
@@ -363,13 +370,37 @@ func (a *AlarmsServer) GetProbableCause(ctx context.Context, request api.GetProb
 	return nil, fmt.Errorf("not implemented")
 }
 
+// AmNotification handles an API request coming from AlertManager with CaaS alerts. This api is used internally.
+// Note: the errors returned can also be view under alertmanager pod logs but also logging here for convenience
 func (a *AlarmsServer) AmNotification(ctx context.Context, request api.AmNotificationRequestObject) (api.AmNotificationResponseObject, error) {
-	// TODO: Implement the logic to handle the AM notification
-	slog.Debug("Received AM notification", "groupLabels", request.Body.GroupLabels)
-	for _, alert := range request.Body.Alerts {
-		slog.Debug("Alert", "fingerprint", alert.Fingerprint, "startsAt", alert.StartsAt, "status", alert.Status)
+	// Audit the table with full list of alerts in the current payload. If missing set them to resolve
+	if err := a.AlarmsRepository.ResolveNotificationIfNotInCurrent(ctx, request.Body); err != nil {
+		msg := "failed to resolve notification that are not present"
+		slog.Error(msg, "error", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+
+	// Get the definition data based on current set of Alert names and managed cluster ID
+	alarmDefinitions, err := a.AlarmsRepository.GetAlarmDefinitions(ctx, request.Body, a.ClusterServer.ClusterIDToResourceTypeID)
+	if err != nil {
+		msg := "failed to get AlarmDefinitions"
+		slog.Error(msg, "error", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
 	}
 
+	// Combine possible definitions with events
+	aerModels := alertmanager.ConvertAmToAlarmEventRecordModels(request.Body, alarmDefinitions, a.ClusterServer.ClusterIDToResourceTypeID)
+
+	// Insert and update AlarmEventRecord
+	if err := a.AlarmsRepository.UpsertAlarmEventRecord(ctx, aerModels); err != nil {
```

Comment:
Why the upsert? Say there is a large number of alert firing at a given time. We will have to update a large number of rows when a single alert is resolved or added. If you loop through the list, you can identify which alerts are Resolved and update them separately. And for new alerts, is the upsert more efficient than querying the DB to identify which are not present?

As followup question about upsert, how can we identify which alerts are New? The sequence on the table is only updated when the alert is resolved or change time has been modified (which should only happen when it is resolved). So, I don't see how we can identify which alerts were inserted by the upsert and not updated. This is important to notifiy (or not notify based on the filter) the users.

Reply:
Sequence is updated when a new row is added + as well as when the conditions are met. So need to insert+update the whole list to allow it auto update as needed. Along with status (like you mentioned)...some of the firing alerts may also new "changed" time or maybe somehow we added the ObjectTypeID that maybe missing previously. Taking one these one at a time is far simpler than to just let upsert do its thing.

Reply:
ok, help me to understand. The sequence starts with 1, when it is added, it goes to the next value, which is 2. But again, how can you identify later that this is a new row? Other rows will also have value 2 when they are not updated. The update happens only when it is resolved because you are not changing alarm_changed_time anywhere.

Reply:
So basically the sequence `alarm_sequence_seq` is global. Meaning every time `nextval('alarm_sequence_seq')` is called (during insertion or update) no one row will have the same value. And then finally this value is used with subscriber's `event_cursor` to determine if that subscriber already been notified or not and update the subscriber cursor. I think the design doc has all the details + example code to explain it better. 

yeah alarm_changed_time is not updated since AM is not sending that value.

Reply:
First of all, thank you for the explanation of how the sequence works. It makes sense now.

Only question that I have now is, why do you expect the AM to send the alarm_changed_time value? The spec clearly says that this value must be set when any value of the AlarmEventRecord has been modified. The AM does not update anything. All is control by us. We modify the record when we acknowledge and/or clear the alert. And now I understand that with the upsert, you want to add another layer of complexity for firing Alerts. You want to introduce for these the possibility to update the AlarmDefinitionID, ResourceID, ResourceTypeID. I don't know if we should add this new level of complexity. But fine. However for all this to work, we need the alarm_changed_time to be modified whenever we touch any of those fields. Otherwise, the sequence won't be updated. It won't be possible to detect a change.

And please, add more comments in the code, especially for the upsert. That you want to be possible to update some fields for current firing alerts.

Reply:
Ok let me add a trigger to keep track of the rows and use that as alarm_changed_time